### PR TITLE
feat: Implement rebalance queue and config option (Fixes #8982)

### LINKED ISSUE #8982: Resource sharing amongst users at the template level
## Summary

I would like to create workflows on behalf of 3rd party users, and I would like those users to be able to share resources equally at the template level via "user-level" parallelism.

## Potential interface

One instantiation of such a feature could look like this (note that I continue to update this as I experiment more with Argo codebase):

The workflow spec has a `ResourceSharingID` that identifies a single participant who will receive the same number of resources as all other participants (basically, a "user").

A workflow could have this ID passed along like this:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: pipeline-
spec:
  entrypoint: sg-wrapper
  resourceSharingID: abc-1233333
  templates:
```

but more realistically, when the workflow is created, it'd be passed in dynamically:

```sh
argo submit -l resourceSharingID=abc123333 workflow.yaml
```

Resource limits could be set in the following way. In `workflow-controller-configmap`:

```yaml
  resourcesAvailable: |
    template-name-1: "4"
    template-name-1: "20"

```

Any templates that are not defined here (or are not used in steps / group) will bypass resource sharing checks, (but will still be subject to all other parallelism checks).

## Details / edge cases:

- If parallelism for a particular step is defined, AND a resource limit is set for it's template, the behavior is undefined but the Workflow will submit.

- Workflows that do not contain resourceSharingID will have no limits set on pods with regards to available resources.

## Notes:

- I do not think such behavior exists (trivially) at the moment, but if I have missed it I'd love to be pointed in the right direction! (I have written a cron that adjusts template-level parallelism on an interval, but this approach seems bad for a variety of reasons)

- I have already set up the codebase in a dev container and have been reading/experimenting with it. I'm beginning to get a feel for what would need to be done - I would be more than happy to work with a mentor to bring this feature back upstream (as I need this feature for a project, I'll probably be implementing either way).

## Use Cases

I am currently building software for a render farm that processes long-running workflows (several hours to days). Access to our resources is bottlenecked by a finite number of provisioned software licenses for our farm.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by robertkotcher**
Signed-off-by: Robert Kotcher <rkotcher+1@gmail.com>

Fixes #8982

* [x] Create the PR as draft .
* [x] Run `make pre-commit -B` to fix codegen and lint problems.
* [x] Sign-off your commits (otherwise the DCO check will fail).
* [x] Use [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) (otherwise the commit message check will fail).
* [x] "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
* [x] Add unit or e2e tests. Say how you tested your changes. If you changed the UI, attach screenshots.
* [x] Github checks are green.
* [x] Once required tests have passed, mark your PR "Ready for review".

### CODE REVIEW

File: config/config.go
Comment (sarabala1979) [reply config/config.go:104]: Instead of having it controller config, Can we have it in the semaphore configmap itself.
The current approach will not support multi-tenant.

**Review by sarabala1979 — CHANGES_REQUESTED** [review]

File: config/config.go
Comment (robertkotcher) [reply config/config.go:104]: Sorry, pinged you on slack but figured it might be best to comment here for better visibility

File: workflow/controller/controller.go
Comment (isubasinghe) [inline workflow/controller/controller.go:354]: Why do you need this "%s-strategy" string formatting? Where does this get populated?

In general I do think it would be nice to move out key generation into functions, it just helps to trace usage better when you can just use the LSP to find references.

File: workflow/sync/rebalanced.go
Comment (isubasinghe) [inline workflow/sync/rebalanced.go:61]: Minor thing but why convert this to a float? Is it not better to cast maxLocksPerUserFloor into an int instead? 
I know we will probably never hit numbers large enough where comparison being imprecise is a worry, but probs nice to use ints regardless.

File: workflow/controller/controller.go
Comment (robertkotcher) [reply workflow/controller/controller.go:354]: The user configures this in config map.

I'm not 100% sure I follow the suggestion. How would the user configure semaphore strategy with a function?

Thx!

File: workflow/sync/rebalanced.go
Comment (robertkotcher) [reply workflow/sync/rebalanced.go:61]: @isubasinghe it's because `maxLocksPerUserFloor` is a float, since Floor returns float64. See line 55.

File: workflow/controller/controller.go
Comment (isubasinghe) [reply workflow/controller/controller.go:354]: Thanks for your clarification, I meant refactor the string formatting out to a function, but if this is the only place it is going to be used, it is probably fine to remain as it is. A comment about it would be great though. 

Thanks

File: workflow/sync/rebalanced.go
Comment (isubasinghe) [reply workflow/sync/rebalanced.go:61]: Yes, I was more suggesting to go with approach on line 67 instead of casting to floats. It honestly is a very minor nit though, probs fine as it is, this is more me just being paranoid

File: workflow/controller/controller.go
Comment (robertkotcher) [reply workflow/controller/controller.go:354]: Oh, thanks! I understand! I will add a comment. I appreciate the review @isubasinghe

Comment (isubasinghe) [reply workflow/controller/controller.go:357]: Thanks for doing that! No worries! I've just approved this

**Review by isubasinghe — APPROVED** [review]
Awesome work! Thanks for your contribution.

File: config/config.go
Comment (robertkotcher) [reply config/config.go]: @sarabala1979 I have implemented this change - do you think I could get your approval on this PR? Or, are there any further changes you'd like me to make? Thanks!

**Review by caelan-io — APPROVED** [review]
Looks like all the requested changes have been made. Thanks, @robertkotcher ! Excited to build on top of this feature and gather more user feedback once it's merged

**Comment by sarabala1979** [comment]
@terrytangyuan Can you review this PR?

**Comment by terrytangyuan** [comment]
> @terrytangyuan Can you review this PR?

Will do

File: docs/proposals/rebalanced-semaphore.md
Comment (terrytangyuan) [inline docs/proposals/rebalanced-semaphore.md:50]: Where is this used?

File: docs/proposals/rebalanced-semaphore.md
Comment (terrytangyuan) [inline docs/proposals/rebalanced-semaphore.md:55]: This does not seem ideal. Why isn't this config dynamically picked up?

**Comment by terrytangyuan** [comment]
Great work on this, but I wonder why https://github.com/argoproj/argo-workflows/issues/8982 is pretty quiet and not popular.

**Comment by robertkotcher** [comment]
> Great work on this, but I wonder why #8982 is pretty quiet and not popular.

It's a fair point. @caelan-io has mentioned that this is something that some of Pipekit customers would benefit from, but I haven't heard any others come out to say it'd be useful. Maybe we can talk more about this in next maintainer meeting. Maybe there's a way to poll or figure out if this would be a useful addition?

**Pull request closed without merging**
